peps/pep-0751.rst

```diff
@@ -0,0 +1,901 @@
+without necessitating running a tool, once again to help with auditing. Finally,
+the format is designed so that viewing a diff of the file is easy by centralizing
+relevant details to minimizing needing to look at other parts of the file.
+
+The file format is also designed to not require a resolver at install time.
+Being able to analyze dependencies in isolation from one another when listed in
+a lock file provides a few benefits. First, it supports auditing by making it
+easy to
+figure out if a certain dependency would be installed for a certain environment
+by looking at each dependency in isolation instead of having to reference other
+parts of the file contextually. It should also lead to faster installs which are
+much more frequent than creating a lock file. Finally, the four tools mentioned
+in the Motivation section either already implement this approach of evaluating
+dependencies in isolation or have suggested they could in
+`Poetry's case <https://discuss.python.org/t/lock-files-again-but-this-time-w-sdists/46593/83>`__.
+
+
+Locking Scenarios
+-----------------
+
+The lock file format is designed to support two locking scenarios. The format
+should also be flexible enough that adding support for other locking scenarios
+is possible via a separate PEP.
+
+
+Per-file Locking
+''''''''''''''''
+
+*Per-file locking* operates under the premise that one wants to install exact
+same files in any matching environment. As such, the lock file specifies the
+requirements for an environment and then matches those environment requirements
+with the files to install. There can be multiple environments specified in a
+single file, each with their own set of files to install. By specifying the
+exact files to install installers avoid performing a resolution to decide what
+to install.
+
+The motivation for this approach to locking is for those who have controlled
+environments that they work with. For instance, if you have specific, controlled
+development and production environments then you can use per-file locking to
+make sure the **same** files are installed in either environment for everyone.
+
+This is similar to what ``pip freeze`` and pip-tools_
+support, but with more strictness of the exact files as well as incorporating
+support to specify the locked files for multiple environments in the same file.
+
+
+Package Locking
+'''''''''''''''
+
+*Package locking* lists the packages and their versions that *may* apply to any
+environment being installed for. The list of packages and their versions are
+evaluated individually and independently from any other packages and versions
+listed in the file. This allows installation to be linear -- read each package
+and version and make an isolated decision as to whether it should be installed.
+This avoids requiring the installer to perform a *resolution* (i.e.
+determine what to install based on what else is to be installed).
```

Comment:
If we need to define what resolving/resolution is, it should be much earlier. 

Also update to match the phrasing in per-file locking, though this does feel repetitive -- is saying that no resolution happens required everywhere it is currently said?
```suggestion
This avoids requiring the installer to perform any resolution to decide what to install.
```

Reply:
I don't view this as defining "resolution" as much as reiterating what's in the Rationale: the format is designed such that installers do not need to go back and reference other parts of the file to make a decision about installing something.